pkg/didcomm/crypto/jwe/authcrypt/authcrypt_test.go

```diff
@@ -86,20 +95,82 @@ func TestEncrypt(t *testing.T) {
 		crypter, e := New(sendEcKey, []*[chacha.KeySize]byte{recipient1Key.pub, recipient2Key.pub, recipient3Key.pub}, XC20P)
 		require.NoError(t, e)
 		require.NotEmpty(t, crypter)
-		enc, e := crypter.Encrypt([]byte("lorem ipsum dolor sit amet"))
+		pld := []byte("lorem ipsum dolor sit amet")
+		enc, e := crypter.Encrypt(pld)
 		require.NoError(t, e)
 		require.NotEmpty(t, enc)
 
 		m, e := prettyPrint(enc)
 		require.NoError(t, e)
+		t.Logf("Encryption with unescaped XC20P: %s", enc)
 		t.Logf("Encryption with XC20P: %s", m)
 
 		dec, e := crypter.Decrypt(enc, recipient1Key.priv)
+		require.NoError(t, e)
+		require.NotEmpty(t, dec)
+		require.EqualValues(t, dec, pld)
+	})
+
+	t.Run("Success test case: Decrypting a message with two Crypter instances to simulate two agents", func(t *testing.T) {
+		crypter, e := New(sendEcKey, []*[chacha.KeySize]byte{recipient1Key.pub, recipient2Key.pub, recipient3Key.pub}, XC20P)
+		require.NoError(t, e)
+		require.NotEmpty(t, crypter)
+		pld := []byte("lorem ipsum dolor sit amet")
+		enc, e := crypter.Encrypt(pld)
+		require.NoError(t, e)
+		require.NotEmpty(t, enc)
+
+		m, e := prettyPrint(enc)
+		require.NoError(t, e)
+		t.Logf("Encryption with unescaped XC20P: %s", enc)
+		t.Logf("Encryption with XC20P: %s", m)
+
+		// now decrypt with recipient3
+		crypter1, e := New(recipient3Key, []*[chacha.KeySize]byte{recipient3Key.pub}, XC20P)
+		require.NoError(t, e)
+		dec, e := crypter1.Decrypt(enc, recipient3Key.priv)
+		require.NoError(t, e)
+		require.NotEmpty(t, dec)
+		require.EqualValues(t, dec, pld)
+
+		// now try decrypting with recipient2
+		crypter2, e := New(recipient2Key, []*[chacha.KeySize]byte{recipient2Key.pub}, XC20P)
+		require.NoError(t, e)
+		dec, e = crypter2.Decrypt(enc, recipient2Key.priv)
+		require.NoError(t, e)
+		require.NotEmpty(t, dec)
+		require.EqualValues(t, dec, pld)
+	})
+
+	t.Run("Failure test case: Decrypting a message with an unauthorized (recipient2) agent", func(t *testing.T) {
+		crypter, e := New(sendEcKey, []*[chacha.KeySize]byte{recipient1Key.pub, recipient3Key.pub}, XC20P)
+		require.NoError(t, e)
+		require.NotEmpty(t, crypter)
+		pld := []byte("lorem ipsum dolor sit amet")
+		enc, e := crypter.Encrypt(pld)
+		require.NoError(t, e)
+		require.NotEmpty(t, enc)
+
+		m, e := prettyPrint(enc)
+		require.NoError(t, e)
+		t.Logf("Encryption with unescaped XC20P: %s", enc)
+		t.Logf("Encryption with XC20P: %s", m)
+
+		// decrypting for recipient 2 (unauthorized)
+		crypter1, e := New(recipient2Key, []*[chacha.KeySize]byte{recipient2Key.pub}, XC20P)
+		require.NoError(t, e)
+		dec, e := crypter1.Decrypt(enc, recipient2Key.priv)
+		require.Error(t, e)
+		require.Empty(t, dec)
+
+		// now try to decrypt with valid recipient who's trying to use another agent's key
+		crypter1, e = New(recipient2Key, []*[chacha.KeySize]byte{recipient2Key.pub}, XC20P)
+		require.NoError(t, e)
+		dec, e = crypter1.Decrypt(enc, recipient1Key.priv)
 		require.Error(t, e)
 		require.Empty(t, dec)
 	})
```

Comment:
How about failure tests for malformed messages? Truncated, missing entries, or with single-byte substitutions to the AAD, for example?

Reply:
yes, more tests are needed.. this was a first push with full end to end flow passing..

Reply:
added more unit tests